### DISCUSSION TITLE
ASPNet Core in Desktop Apps

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#85095] [Q&A] [answered]
I have a desktop app that needs its own internal HTTPServer.
Do I need to create a whole new project that references a ```Microsoft.NET.Sdk.Web``` project or are there nuget packages I can reference to bring in all the AspNetCore stuff?

I seems like all/most of the ASPNet Core packages have not been updated in years.

**Comment by gfoidl** [comment, marked as answer]
See [Use ASP.NET Core APIs in a class library](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/target-aspnetcore?view=aspnetcore-7.0&tabs=visual-studio).

---

>  a desktop app that needs its own internal HTTPServer.

Out of interest, can you tell why you need this approach?

**Comment by TonyValenti** [comment]
Thank you!  This is exactly what I was looking for.
In our app, we spin up a micro HTTP server that is used to handle OAUTH authentication requests for our desktop application.

**Comment by TonyValenti** [comment]
@gfoidl - Do you happen to know how to get an IWebHost out of an IHost?

I have the following code I am trying to convert from IWebHostBuilder to IHostBuilder.

```
var MyHost = Builder()
                .ConfigureWebHost(x => {
                    x.UseKestrel(o => {
                        o.Listen(System.Net.IPAddress.Loopback, 0);
                    });
                })
                .Build()
                ;

            var MyHostTask = MyHost.RunAsync(TokenSource.Token);

            //_________This doesnt work.  What should I do?
            var WebHost = MyHost.Services.GetRequiredService<IWebHost>();

            var Urls = (
                from x in (WebHost.ServerFeatures.Get<IServerAddressesFeature>()?.Addresses).EmptyIfNull()
                select x
                ).ToLinkedList();
```

**Reply by gfoidl** [reply]
Grab the `IServer` and take the features from this one.

With `IHostBuilder` / `IHost`:
```c#
IHostBuilder hostBuilder = Host.CreateDefaultBuilder(args);
IHost host = hostBuilder.Build();

IServer server = host.Services.GetRequiredService<IServer>();
IServerAddressesFeature? urls = server.Features.Get<IServerAddressesFeature>();
// ...
host.Run();
```

Or with `WebApplicationBuilder` / `WebApplication`:
```c#
WebApplicationBuilder builder = WebApplication.CreateSlimBuilder(args);
WebApplication app = builder.Build();

IServer server = app.Services.GetRequiredService<IServer>();
IServerAddressesFeature? urls = server.Features.Get<IServerAddressesFeature>();
// ...
app.Run();
```

---

> `.ToLinkedList();`

Why use a linked list here, and not a simple `List<T>`? Would be cheaper, as the contents of the list are populated here.

**Reply by TonyValenti** [reply]
Thanks so much!

Good catch on the ToLinkedList!  Totally not needed.
